### PR TITLE
Fix FileManager.createFile on Windows

### PR DESCRIPTION
**Pull request by gmittert**
- rename doesn't work on Windows if the target already exists, instead
  use ReplaceFile/MoveFile depending on if the target exists
- open() doesn't properly set sharing permissions, use CreateFile
  instead
- fds must be closed before attempting to move their target file on
  Windows

```swift
import Foundation
print(FileManager.default.createFile(atPath: "Foo.txt", contents: Data(), attributes: nil))
print(FileManager.default.createFile(atPath: "Foo.txt", contents: Data(), attributes: nil))
```
didn't work before, now it does.

### CODE REVIEW

File: Foundation/NSPathUtilities.swift [hunk 1]
```diff
@@ -663,7 +663,19 @@ internal func _NSCreateTemporaryFile(_ filePath: String) throws -> (Int32, Strin
...
-    let fd = open(pathResult, _O_CREAT)
+    guard let h = CreateFileW(buf,
+                              GENERIC_READ | DWORD(GENERIC_WRITE),
+                              DWORD(FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE),
+                              nil,
+                              DWORD(OPEN_EXISTING),
+                              DWORD(FILE_ATTRIBUTE_NORMAL),
+                              nil),
+          h != INVALID_HANDLE_VALUE else {
+      throw _NSErrorWithErrno(Int32(GetLastError()), reading: false,
```
Comment (compnerd) [inline Foundation/NSPathUtilities.swift]: Please use `_NSErrorWithWindowsError`

File: Foundation/NSPathUtilities.swift [hunk 2]
```diff
@@ -678,10 +690,24 @@ internal func _NSCreateTemporaryFile(_ filePath: String) throws -> (Int32, Strin
 }
 
 internal func _NSCleanupTemporaryFile(_ auxFilePath: String, _ filePath: String) throws  {
+#if os(Windows)
+    try auxFilePath.withCString(encodedAs: UTF16.self) { fromPath in
+        try filePath.withCString(encodedAs: UTF16.self) { toPath in
+            let res = PathFileExistsW(toPath)
+                ? ReplaceFileW(toPath, fromPath, nil, 0, nil, nil)
+                : MoveFileW(fromPath, toPath)
```
Comment (compnerd) [inline Foundation/NSPathUtilities.swift]: Ugh, this is racy.  The path can be created in between `PathFileExistsW` and `MoveFileW`.

Comment (gmittert) [reply Foundation/NSPathUtilities.swift]: I've changed this to use CopyFileW and then just delete the temp file.

**Comment by gmittert** [comment]
@swift-ci please test

File: Foundation/NSData.swift [hunk 3]
```diff
@@ -453,13 +453,27 @@ open class NSData : NSObject, NSCopying, NSMutableCopying, NSSecureCoding {
...
+#else
+            let shouldCleanUpFD = true
+#endif
+            let fh = FileHandle(fileDescriptor: newFD, closeOnDealloc: shouldCleanUpFD)
             do {
                 try doWrite(fh)
+#if os(Windows)
+                // Moving a file (via _NSCleanupTemporaryFile) requires that
+                // there be no open handles to the file
+                _close(newFD)
```
Comment (compnerd) [inline Foundation/NSData.swift]: I get that we need to close the handle, but I think that we can share more of the path by using `FileHandle.closeFile()` on Windows explicitly.

**Comment by gmittert** [comment]
@swift-ci please test

**Comment by gmittert** [comment]
@swift-ci please test

**Comment by gmittert** [comment]
@swift-ci please test and merge

**Comment by gmittert** [comment]
@swift-ci please test and merge

**Comment by gmittert** [comment]
@swift-ci  please test and merge

**Comment by gmittert** [comment]
@swift-ci please test and merge

**Pull request merged**